### DISCUSSION TITLE
Refactor 1: Config

### DISCUSSION THREAD

**Discussion by TornaxO7** [leftwm/leftwm#886] [Ideas]
Yes yes yes, we've already talked about this one on Discord but I'm still not satisfied with the answers yet.

I'm still convinced that having a struct is better for a config then a trait. To not reapet some old arguments, here are some arguments from discord with an opinion from me:

> the struct forces you to a specific format. a trait allows you to define your own struct

A trait forces you to a specific format as well, since the return type of the trait equals the type of the config struct, so it's like creating the config struct where each attribute comes from a function (see example below).
Also, if you want to have a diffrent struct-format with different attributes, a simple `impl From<MyConfig> for Config` should be able do the job for this.

> for example, in my case i dont use a config format at all, i just define a struct that implements the trait and i return the value immediately (because i dont want a config file)

It's even easier in my opinion to the values into the struct:
```rust
Config {
    val1: vec![bla bla],
    .. Config::default(),
}
```
or if you want to do more magic for an attribute:
```rust
Config {
    val1: get_val1(),
    .. Config::default(),
}

fn get_val1() -> <type> {
    // ...
}
```

@cecton 

## Other reasons for a struct
- There's only one config for leftwm and this is fixed to the types of the return value of the trait or for a struct: The types of the attribute. Even if you have millions of different structs which implement the config trait, those can be still easily converted into the config struct, since the only relevant information is the return value of the trait (and they shouldn't change during runtime).
- Also when writing code, we wouldn't need to add the brackets if we want to access a config value which means less to type and less to read

> I don't have much opinion on this ^_^ as long as we don't re-introduce struct that has Serialize and Deserialize

may I ask here what your cons for a config struct with serialize and deserialize are? @cecton Maybe I'm just a bit too blind too see them currently ._.

**Comment by VuiMuich** [comment]
Additionally to the reasons already discussed, at this point I don't see a refactor of the config reasonably, as we want to change the feature set of the config by a lot (that was one major reason to move away from toml in the first place).
But of course this discussion can go on in the meantime on a theoretical basis and might be taken into account for, when introducing the new features.

**Comment by cecton** [comment]
I should have answered here too :sweat: 

So why the trait and not a struct?

It was originally a struct that implements Serialize and Deserialize but that was putting a lot of constraints. For example, I wanted to add arguments to the `Command` enum but because it was use "as is" in the configuration, `Command` had to implement Serialize and Deserialize and tuple arguments in enum variants are not supported. So I wanted to make a better separation between what constitutes a configuration file and the configuration object used by leftwm.

In parallel, I wanted to make a separation between the core functionality of leftwm (being a tiling manager and that's it) and the user functionalities (reading configuration, theme, programming behavior, etc...) in order to allow leftwm to be used as a library and not forcibly as a standalone program.

As a library, you can use it in your project and do whatever you want. For example, if you would want to connect to a remote database, you can, if you want to control a session remotely, you can, control with dbus or a unix socket, ... it doesn't matter what you want to do because if the API is done well enough, you can extend with any functionality you want.

In my particular use case I want to have a standalone binary for my own personal desktop manager and another one for a gaming session (where I would run Steam). In my personal desktop manager I like to have a window on the left and one on the right but on my gaming session everything needs to be fullscreen. Some other behaviors change like that (no topbar, no border, etc...).

I also like to have a compiled binary instead of binary + configuration files. A bit like XMonad. (Maybe a subject for a bigger discussion but in this case I'm aiming to allow the user to choose: they can use it as a lib (full customization) or as a bin (includes a set of useful features like loading configuration, communicating with leftwm through the socket (not yet in the bin but should be xD), themes, etc...))

Now to come back to the question of why a trait and not another struct: because if you extend the configuration "simply", then a struct is sufficient. But if you want to go deeper and put actual code and logic, then it's best to have a trait where you can actually write code. Here is a sample from my own configuration (far from finished):

```rust
    fn setup_predefined_window(&self, window: &mut Window) -> bool {
        match (window.res_class.as_deref(), window.res_name.as_deref()) {
            // TODO apparently the 2 classes from xprop are in res_class and res_name. It doesn't
            //      seem to be an array but more like a tuple that identify a specific window's
            //      program and role
            (Some("firefox"), Some("Toolkit")) => {
                window.set_floating(true);
                true
            }
            (Some("gui"), Some("gui")) => {
                window.set_floating(true);
                true
            }
            (class, name) => {
                log::info!("setup_predefined_window: {:?} {:?}", class, name);
                false
            }
        }
    }
```

The code is pretty simple here but I could also store a state in my configuration object and code some pretty complicated behavior if I want.

Now, that being said, there is quite a gap between what I think I would do and it would look like compared to what actually has been done and currently look like. So I understand the current state looks over-engineered because we don't see much the benefit. As @VuiMuich suggest I should share my config or write a blog post about it.

(Note that it is also possible to add code to a struct using boxes:

```rust
struct Config {
    setup_predefined_window: Box<dyn Fn(&mut Window) -> bool>,
}
```

Which would allow simplifying the generic arguments but it will look ugly in some other ways...)

**Reply by cecton** [reply]
(I don't want anyone to feel blocked by me so if you have ideas and want to move ahead please don't hesitate!)

**Reply by TornaxO7** [reply]
Thank you first of all for yours answer! I didn't fully read it yet, but I'm happy for writing your arguments down!

> As @VuiMuich suggest I should share my config or write a blog post about it.

That would be really interesting indeed to see, if a config trait is more suitable here!
